keps/sig-storage/2485-read-write-once-pod-pv-access-mode/README.md

```diff
@@ -0,0 +1,972 @@
+mode to PersistentVolumes and PersistentVolumeClaims. This change will require
+adding a feature gate and updating validation logic to accept this new access
+mode type if the feature gate is enabled.
+
+```golang
+       // can be mounted read/write mode to exactly 1 pod
+       ReadWriteOncePod PersistentVolumeAccessMode = "ReadWriteOncePod"
+```
+
+This access mode will be enforced in two places:
+
+- First is at the time a volume is mounted for filesystem devices, and at the
+  time a volume is mapped for block devices. During the mount operation, kubelet
+  will check the actual state of the world to determine if the volume is already
+  in-use by another pod. If it is, kubelet will fail mounting with an
+  appropriate error message.
+- As an additional precaution, this will also be enforced during scheduling.
+  When scheduling a pod, if another pod is found using the same PVC and the PVC
+  uses ReadWriteOncePod, then scheduling will fail and the pod will be
+  considered unresolvable.
+
+### CSI Specification Changes, Volume Capabilities
+
+In the CSI spec we should add two new access modes that explicitly state the
+number of writers on a single node.
+
+```protobuf
+      // Can only be published once as read/write at a single worklad on
+      // a single node, at any given time.
+      SINGLE_NODE_SINGLE_WRITER = 6;
+
+      // Can be published as read/write at multiple workloads on a
+      // single node simultaneously.
+      SINGLE_NODE_MULTI_WRITER = 7;
+```
+
+These access modes are modeled after the existing `MULTI_NODE_SINGLE_WRITER` and
+`MULTI_NODE_MULTI_WRITER` access modes. The reason for making this distinction
+is because the `SINGLE_NODE_WRITER` volume capability has conflicting
+definitions (see the [motivation](#motivation) section for context).
+
+For CSI clients, the new ReadWriteOncePod Kubernetes access mode will map to the
+`SINGLE_NODE_SINGLE_WRITER` volume capability access mode in the CSI spec.
+
+For the ReadWriteOnce access mode, the value it maps to depends on the CSI
+driver. If the CSI driver supports the `SINGLE_NODE_SINGLE_WRITER` access mode,
+then ReadWriteOnce will map to that value. If the CSI driver does not support
+the `SINGLE_NODE_SINGLE_WRITER` access mode, then ReadWriteOnce will map to
+`SINGLE_NODE_WRITER` to preserve backwards compatibility. In order to determine
+which mapping to use, both the controller and node services should have
+capability bits for this access mode.
+
+```protobuf
+      // Indicates the SP supports the SINGLE_NODE_MULTI_WRITER access
+      // mode.
+      SINGLE_NODE_MULTI_WRITER = 13;
+```
+
+Put more succinctly:
+
+|                  | Driver Supports `SINGLE_NODE_*_WRITER` | Driver Does Not Support `SINGLE_NODE_*_WRITER` |
+|------------------|----------------------------------------|------------------------------------------------|
+| ReadWriteOncePod | SINGLE_NODE_SINGLE_WRITER              | FAILED_PRECONDITION                            |
```

Comment:
What returns `FAILED_PRECONDITION`? One can use `ReadWriteOncePod` volume with a storage backend that supports only `SINGLE_NODE_WRITER` without any issues - scheduler & kubelet changes should be enough to enforce the correct behavior in most cases.

Reply:
Yes this wrong, `FAILED_PRECONDITION` won't be returned. It looks like the external-provisioner will [map this access mode to `nil`](https://github.com/kubernetes-csi/external-provisioner/blob/v2.2.0/pkg/controller/controller.go#L469). The behavior depends on the CSI driver, for example the [EBS CSI driver will error if it receives an unsupported access mode](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/pkg/driver/controller.go#L117-L122).